sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/SaveAsHiveFile.scala

```diff
@@ -124,11 +153,22 @@ private[hive] trait SaveAsHiveFile extends DataWritingCommand {
     val hiveVersion = externalCatalog.unwrapped.asInstanceOf[HiveExternalCatalog].client.version
     val stagingDir = hadoopConf.get("hive.exec.stagingdir", ".hive-staging")
     val scratchDir = hadoopConf.get("hive.exec.scratchdir", "/tmp/hive")
+    logDebug(s"path '${path.toString}', staging dir '$stagingDir', " +
+      s"scratch dir '$scratchDir' are used")
 
     if (hiveVersionsUsingOldExternalTempPath.contains(hiveVersion)) {
       oldVersionExternalTempPath(path, hadoopConf, scratchDir)
     } else if (hiveVersionsUsingNewExternalTempPath.contains(hiveVersion)) {
-      newVersionExternalTempPath(path, hadoopConf, stagingDir)
+      // HIVE-14270: Write temporary data to HDFS when doing inserts on tables located on S3
+      // Copied from Context.java#getTempDirForPath of Hive 2.3.
+      if (supportSchemeToUseNonBlobStore(path)) {
+        // Hive sets session_path as HDFS_SESSION_PATH_KEY(_hive.hdfs.session.path) in hive config
+        val sessionScratchDir = externalCatalog.unwrapped.asInstanceOf[HiveExternalCatalog]
+          .client.getConf("_hive.hdfs.session.path", "")
```

Comment:
Nit: Should we define a constant for hive.hdfs.session.path ?

Reply:
Added.